### PR TITLE
[SPARK-35827][SQL] Show proper error message when update column types to year-month/day-time interval

### PR DESCRIPTION
**Pull request by sarutak**
### What changes were proposed in this pull request?

This PR fixes error message shown when changing a column type to year-month/day-time interval type is attempted.

### Why are the changes needed?

It's for consistent behavior.
Updating column types to interval types are prohibited for V2 source tables.
So, if we attempt to update the type of a column to the conventional interval type, an error message like `Error in query: Cannot update <table> field <column> to interval type;`.

But, for year-month/day-time interval types, another error message like `Error in query: Cannot update <table> field <column>:<type> cannot be cast to interval year;`.

You can reproduce with the following procedure.
```
$ bin/spark-sql
spark-sql> SET spark.sql.catalog.mycatalog=<a catalog implementation class>;
spark-sql> CREATE TABLE mycatalog.t1(c1 int) USING <V2 datasource implementation class>;
spark-sql> ALTER TABLE mycatalog.t1 ALTER COLUMN c1 TYPE interval year to month;
```

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Modified an existing test.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 1]
```diff
@@ -533,7 +533,8 @@ trait CheckAnalysis extends PredicateHelper with LookupCatalog {
                   case u: UserDefinedType[_] =>
                     alter.failAnalysis(s"Cannot update ${table.name} field $fieldName type: " +
                       s"update a UserDefinedType[${u.sql}] by updating its fields")
-                  case _: CalendarIntervalType =>
+                  case _: CalendarIntervalType | _: YearMonthIntervalType |
+                       _: DayTimeIntervalType =>
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:537]: Hm, so we should fix all these kind of blacklisting .. while this fix is good, I wonder if we can sweep and fix in batch.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
cc @gengliangwang and @MaxGekk FYI

File: sql/core/src/test/scala/org/apache/spark/sql/connector/AlterTableTests.scala [hunk 2]
```diff
@@ -403,8 +403,24 @@ trait AlterTableTests extends SharedSparkSession {
...
+        "interval day",
+        "interval hour",
+        "interval minute",
+        "interval second",
+        "interval day to second",
+        "interval day to minute",
+        "interval day to hour",
+        "interval hour to second",
+        "interval hour to minute",
+        "interval minute to second").foreach { case dtype =>
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/AlterTableTests.scala:420]: You can get all year-month and day-time interval types from `DataTypeTestUtils`, see `dayTimeIntervalTypes` and `yearMonthIntervalTypes`. Please, take them, and call `typeName`.

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/connector/AlterTableTests.scala (hunk 2 above)
Comment (sarutak) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/AlterTableTests.scala:420]: Ah, O.K, I'll replace them. Thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala (hunk 1 above)
Comment (sarutak) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:537]: Actually, I have found another places where `DayTimeINtervalType` and `YearMonthIntervalType` are not considered.
In terms of blacklisting, I found this part only (if I didn't miss something).
The rest of places we need to consider is as follows.

* https://github.com/apache/spark/blob/4f51e0045e2754035e304979182c0f583c65f174/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateUnsafeProjection.scala#L42
* https://github.com/apache/spark/blob/4f51e0045e2754035e304979182c0f583c65f174/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala#L155
* https://github.com/apache/spark/blob/4f51e0045e2754035e304979182c0f583c65f174/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala#L191
* https://github.com/apache/spark/blob/4f51e0045e2754035e304979182c0f583c65f174/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ArrayBasedMapBuilder.scala#L37
* https://github.com/apache/spark/blob/4f51e0045e2754035e304979182c0f583c65f174/sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcHadoopFsRelationSuite.scala#L40
* https://github.com/apache/spark/blob/master/sql/hive/src/test/scala/org/apache/spark/sql/sources/ParquetHadoopFsRelationSuite.scala#L41
* https://github.com/apache/spark/blob/4f51e0045e2754035e304979182c0f583c65f174/sql/hive/src/test/scala/org/apache/spark/sql/sources/JsonHadoopFsRelationSuite.scala#L35
* https://github.com/apache/spark/blob/4f51e0045e2754035e304979182c0f583c65f174/sql/hive/src/test/scala/org/apache/spark/sql/sources/SimpleTextHadoopFsRelationSuite.scala#L36

I think, we need to fix them with beyond just blacklisting and they need to be more carefully considered while this part can be simply fixed.

**Review by MaxGekk — APPROVED** [review]
+1, LGTM. Merging to master.
Thank you, @sarutak and @HyukjinKwon for review.

**Pull request closed without merging**